text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+[`is_authorized` method](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L300).
+
+While the majority of requests likely operate on a single resource and can be
+handled efficiently by the current implementation, some requests (especially
+`list` requests) may need to check authorization against multiple resources, and
+it is inefficient to do a full round-trip for every request. Similarly, a
+higher-level operation may involve multiple lower-level actions on multiple
+resources.
+
+In naive comparison testing with the
+[k9securityio/cedar-py](https://github.com/k9securityio/cedar-py)
+`is_authorized` implementation against a
+[very naive `is_authorized_batch` implementation](https://github.com/glb/cedar-py/commit/98bf3e1fa2e90bb6f7326cbd846a3293fdbd960b),
+I observed speedups of at least 5x for even simple cases, as documented in
+[k9securityio/cedar-py#13](https://github.com/k9securityio/cedar-py/issues/13).
+
+@skuenzli also
+[reported](https://github.com/k9securityio/cedar-py/issues/13#issuecomment-1629790294)
+that:
+
+> The time observed in Python is consistently 2 orders of magnitude greater than
+> the authorizer.is_authorized call. These particular requests use a simple
+> policy, a simple schema, and ~170 entities.
+
+The initial thought was to simply create an `is_authorized_batch` API for the
+`cedar-py` Python binding, but in discussion in the Cedar Policy Language Slack
+there was support from @aaronjeline and @andrewmwells-amazon for exposing this
+API in Cedar core so that it could be picked up more easily by other language
+bindings, as they expect those bindings to have similar needs and performance
+concerns.
+
+## Detailed design
+
+### Accept a batch of requests with the current shape
+
+This approach creates a new `is_authorized_batch` function, similar to the
+`is_authorized` function, that accepts a list of `Request` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L2006))
+and returns a list of `Response` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L338)),
+which might have an implementation something like this:
+
+```rust
+impl Authorizer {
+    /// Returns a vector of authorization responses for the input vector of requests
+    /// with respect to the given `PolicySet` and `Entities`.
+    ///
+    /// The order of responses matches the order of requests.
+    pub fn is_authorized_batch(
+        &self,
+        requests: &Vec<Request>,
+        policy_set: &PolicySet,
+        entities: &Entities,
+    ) -> Vec<Response> {
+        let mut responses: Vec<Response> = vec![];
+
+        for request in requests {
+            let response = self.is_authorized(request, policy_set, entities);
+            responses.push(response);
+        }
+
+        responses
+    }
+}
+```
```

Comment:
Copying conversation from https://github.com/cedar-policy/rfcs/pull/14#discussion_r1261468425 here.
I suggest we implement the API you propose such that the Evaluator is only constructed once (which can be substantially more efficient if there are many complex entities). I think we'll have to do something like that inevitably, so I don't find `drawback 1` very concerning.

Reply:
Thanks @andrewmwells-amazon ... I've done a bit of poking at the `Evaluator` and it looks like the majority of the work done in constructing an `Evaluator` is in constructing the `EntityAttrValues`:

https://github.com/cedar-policy/cedar/blob/55c786a9a9f48c309e3373d4ec253f566fcc9f52/cedar-policy-core/src/evaluator.rs#L197-L224

(the `RestrictedEvaluator` created in line 214 seems to just store a reference to the extensions)

https://github.com/cedar-policy/cedar/blob/55c786a9a9f48c309e3373d4ec253f566fcc9f52/cedar-policy-core/src/evaluator.rs#L79-L81

but I'm not sure how much work is done by the call to `partial_interpret` on the context.

I'm wondering if it would make sense to extract the `EntityAttrValues::new` call a level (or two) so that the `EntityAttrValues` is re-used by `is_authorized_batch` but otherwise a new `Evaluator` is still constructed for each request.

Thoughts?

Reply:
Yeah, I think that makes sense. If `is_authorized_batch` takes in an Evaluator or an `EntityAttrValues` then a user with a large entity store and a stream of non-batched requests could call `is_authorized_batch` with a batch size of 1 and not have to pay for repeatedly constructing an `Evaluator` (where building the `EntityAttrValues` is the expensive part).

Reply:
@andrewmwells-amazon streaming is an interesting take. I was thinking that `is_authorized_batch` would construct the `EntityAttrValues` once and then iterate over the requests in the batch and assuming that the provided `Entities` was either a subset of the full store or a dynamic store like in #16 . To me the batch version fits pretty well with the motivating use cases that I had in mind. Your take seems to fit well with @cdisselkoen 's [proposed alternative](https://github.com/cedar-policy/rfcs/pull/14#discussion_r1261277984), and brings back the question I had about lifetimes: who would own the `EntityAttrValues` and make sure that it was cleaned up when it was no longer needed? (recalling that another of the motivating factors here was improving performance for batches of requests passed from a binding written in another language). How does that sort of thing usually get handled in Rust?

Reply:
I agree we want to evaluate the attr values once and reuse across many calls to `is_authorized`.  See https://github.com/cedar-policy/cedar/issues/227.  I believe the solution we'll go for there is that the `Entities` object that comes into the main `is_authorized` API will under-the-hood carry evaluated attr values already -- this means that the work of evaluating the values moves from `is_authorized` to the construction of the `Entities` itself, and thus can be amortized across many `is_authorized` calls.  Given that, the question merely becomes about the lifetime of the `Entities` object.

I'm not sure how lifetimes are normally handled across FFI boundaries, but one option could be to use `Rc` (ref-counted pointer) on the Rust side, increment the refcount when returning the object to the foreign language, and provide a new API call that the foreign language could use to dispose of the object (on the Rust side, decrementing its refcount).  Concretely, in the Python case, the "opaque reference" returned to Python would be something like an `Rc<Entities>` where we manually increase the refcount by one as part of the FFI interface; and there would be a Python function `dispose()` or something, where calling `dispose()` on that opaque reference would call back to Rust where we'd manually decrement the refcount.  But again, I don't know what the typical patterns are for lifetimes across FFI calls.

Reply:
Thanks @cdisselkoen . It seems to me that changing the behaviour so that the work [currently being done to construct](https://github.com/cedar-policy/cedar/blob/55c786a9a9f48c309e3373d4ec253f566fcc9f52/cedar-policy-core/src/evaluator.rs#L203) `EntityAttrValues` could be done earlier (or lazily) would indeed reduce the motivation for an `is_authorized_batch` in the Cedar API level. Language bindings could still choose between the opaque reference solution and a batch operation depending on the behaviour that their users needed. For example, the use cases that I'm contemplating would not particularly benefit from a long-lived reference, but others might definitely want such a thing.